eclair-core/src/main/scala/fr/acinq/eclair/payment/MultiPartPaymentHandler.scala

```diff
@@ -0,0 +1,143 @@
+
+  startWith(WAITING_FOR_HTLC, WaitingForHtlc(0 msat, Nil))
+
+  when(WAITING_FOR_HTLC) {
+    case Event(PaymentTimeout, d: WaitingForHtlc) =>
+      goto(PAYMENT_FAILED) using PaymentFailed(d.paidAmount, wire.PaymentTimeout, d.parts)
+
+    case Event(MultiPartHtlc(totalAmount2, htlc), d: WaitingForHtlc) =>
+      require(htlc.paymentHash == paymentHash, s"invalid payment hash (expected $paymentHash, received ${htlc.paymentHash}")
+      val pp = PartialPayment(htlc.amountMsat, htlc.channelId)
+      if (totalAmount != totalAmount2) {
+        log.warning(s"multi-part payment total amount mismatch: previously $totalAmount, now $totalAmount2")
+        goto(PAYMENT_FAILED) using PaymentFailed(d.paidAmount, IncorrectOrUnknownPaymentDetails(totalAmount2, nodeParams.currentBlockHeight), PendingPayment(htlc.id, pp, sender) :: d.parts)
+      } else if (htlc.amountMsat + d.paidAmount >= totalAmount) {
+        goto(PAYMENT_SUCCEEDED) using PaymentSucceeded(htlc.amountMsat + d.paidAmount, PendingPayment(htlc.id, pp, sender) :: d.parts)
+      } else {
+        stay using d.copy(paidAmount = d.paidAmount + htlc.amountMsat, parts = PendingPayment(htlc.id, pp, sender) :: d.parts)
+      }
+  }
+
+  when(PAYMENT_SUCCEEDED) {
+    // A sender should not send us additional htlcs for that payment once we've already reached the total amount.
+    // However if that happens the only rational choice is to fulfill it, because the pre-image has been released so
+    // intermediate nodes will be able to fulfill that htlc anyway. This is a harmless spec violation.
+    case Event(MultiPartHtlc(_, htlc), _) =>
+      require(htlc.paymentHash == paymentHash, s"invalid payment hash (expected $paymentHash, received ${htlc.paymentHash}")
+      log.info(s"received extraneous htlc for payment hash $paymentHash")
+      parent ! ExtraHtlcReceived(paymentHash, PendingPayment(htlc.id, PartialPayment(htlc.amountMsat, htlc.channelId), sender), None)
+      stay
+
+    case Event("ok", _) => stay
+  }
+
+  when(PAYMENT_FAILED) {
+    // If we receive htlcs after the multi-part payment has expired, we must fail them.
+    // The LocalPaymentHandler will create a new instance of MultiPartPaymentHandler to handle a new attempt.
+    case Event(MultiPartHtlc(_, htlc), PaymentFailed(_, failure, _)) =>
+      require(htlc.paymentHash == paymentHash, s"invalid payment hash (expected $paymentHash, received ${htlc.paymentHash}")
+      parent ! ExtraHtlcReceived(paymentHash, PendingPayment(htlc.id, PartialPayment(htlc.amountMsat, htlc.channelId), sender), Some(failure))
+      stay
+
+    case Event("ok", _) => stay
+  }
+
+  onTransition {
+    case _ -> PAYMENT_SUCCEEDED =>
+      cancelTimer(PaymentTimeout.toString)
+      nextStateData match {
+        case PaymentSucceeded(_, parts) =>
+          // We expect the parent actor to send us a PoisonPill after receiving this message.
+          parent ! MultiPartHtlcSucceeded(paymentHash, parts.reverse)
```

Comment:
Using a `Queue` for `parts` would allow an O(1) append and remove the need for `parts.reverse`.

Reply:
Given the number of HTLC in a set I think this is really overkill :)
We need proper perf testing for that but I'd even argue that it's likely that for small arrays like this a `Queue` will end-up being _more_ costly.